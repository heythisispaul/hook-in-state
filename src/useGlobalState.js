// eslint-disable-next-line no-unused-vars
import React, { createContext, useReducer } from 'react';
import { produce } from 'immer';
import GlobalContext, { GlobalStateProvider } from './context';
import { getInStateBuilder } from './helpers';

export const useSetupGlobalState = (providedActions = {}, initialState = {}) => {
  const getNextState = (state = initialState, action) => (
    produce(state, (draft) => {
      const { type, value } = action;
      const matchingReducer = providedActions[type];
      if (matchingReducer) {
        matchingReducer(draft, value);
      }
    })
  );
  const [state, dispatch] = useReducer(getNextState, initialState);
  const actions = {};
  Object.entries(providedActions).forEach((currentAction) => {
    const [name] = currentAction;
    const type = name.toString();
    actions[type] = (value) => dispatch({ type, value });
  });

  const getInState = getInStateBuilder(state);
  const contextValues = { state, dispatch, actions, getInState };

  const getGlobalStateProvider = GlobalStateProvider(contextValues);

  return {
    ...contextValues,
    GlobalStateProvider: getGlobalStateProvider,
  };
};

const useGlobalState = () => {
  // not desctructuring for easier mocking
  const globalContextValue = React.useContext(GlobalContext);
  return globalContextValue;
};

export default useGlobalState;
