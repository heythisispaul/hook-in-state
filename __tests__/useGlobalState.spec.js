/* eslint-disable no-unused-vars */
import React from 'react';
import { renderHook, act } from '@testing-library/react-hooks';
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import useGlobalState, { useSetupGlobalState } from '../src/useGlobalState';
import GlobalState from '../src/context';
import { createMockState } from './testHelpers/createMockState';

configure({ adapter: new Adapter() });

describe('useGlobalState', () => {
  describe('useSetupGlobalState', () => {
    const mockProvidedActions = {
      increment: (state) => state.count++,
      updateExample: (state, value) => state.anotherExample = value,
    };

    it('maps the provided actions to dispatches', () => {
      expect.assertions(2);
      const state = createMockState();
      const { result } = renderHook(() => useSetupGlobalState(mockProvidedActions, state));
      expect(result.current.actions).toMatchSnapshot();
      act(() => {
        result.current.actions.increment();
      });
      expect(result.current.state.count).toBe(1);
    });

    it('provides actions that dispatch with their provided argument', () => {
      const state = createMockState();
      const { result } = renderHook(() => useSetupGlobalState(mockProvidedActions, state));
      const exampleText = 'example updated!';
      act(() => {
        result.current.actions.updateExample(exampleText);
      });
      expect(result.current.getInState('anotherExample')).toEqual(exampleText);
    });

    it('calls the provided fuction when using a custom dispatch', () => {
      const state = createMockState();
      const { result } = renderHook(() => useSetupGlobalState(mockProvidedActions, state));
      const customDispatchText = 'updated from custom dispatch!';
      act(() => {
        result.current.dispatch(customDispatchText, (state, value) => state.anotherExample = value);
      });
      expect(result.current.state.anotherExample).toEqual(customDispatchText);
    });

    it('uses an empt object for state if none provided', () => {
      const { result } = renderHook(() => useSetupGlobalState(mockProvidedActions));
      expect(result.current.state).toEqual({});
    });

    it('provides the context providing component and it renders its children', () => {
      const state = createMockState();
      const { result } = renderHook(() => useSetupGlobalState(mockProvidedActions, state));
      const { GlobalStateProvider } = result.current;
      const render = mount(
        <GlobalStateProvider>
          <div>I'll render!</div>
          <div>So will I!</div>
        </GlobalStateProvider>
      );
      expect(render.html()).toMatchSnapshot();
    });

    describe('useGlobalState', () => {
      it('should provide the global context', () => {
        const useContextMock = React.useContext = jest.fn();
        renderHook(() => useGlobalState());
        expect(useContextMock).toHaveBeenCalled();
      });
    });
  });
});
