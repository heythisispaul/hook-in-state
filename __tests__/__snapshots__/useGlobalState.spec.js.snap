// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useGlobalState useSetupGlobalState maps the provided actions to dispatches 1`] = `
Object {
  "increment": [Function],
  "updateExample": [Function],
}
`;

exports[`useGlobalState useSetupGlobalState provides the context providing component and it renders its children 1`] = `"<div>I'll render!</div><div>So will I!</div>"`;
